### PR TITLE
CI, TST: early draft of functional testing

### PR DESCRIPTION
**Pull request by tylerjereddy**
* this adds end-to-end functional CI testing from
the darshan runtime monitoring MPI IO activity
right through to the Python bindings checking
for the expected result; you can run this locally (containerized)
with the `act` tool with i.e.,
`act -j test_runtime`

* when cherry-picked into gh-751, this reproduces
the exact error I was hoping CI could catch as
described here:
https://github.com/darshan-hpc/darshan/pull/751#issuecomment-1150197393

* otherwise, running this locally produces the expected failure
to capture `H5D` events when reading with `h5py`, with
only the write, but not the read event going through
the expected API path until Shane's PR (gh-751) is
ready

```
| >           assert h5d_df_counters["H5D_OPENS"].to_numpy() == n_ranks * 2
| E           assert array([1]) == (1 * 2)
| E            +  where array([1]) = <bound method IndexOpsMixin.to_numpy of 0    1\nName: H5D_OPENS, dtype: int64>()
| E            +    where <bound method IndexOpsMixin.to_numpy of 0    1\nName: H5D_OPENS, dtype: int64> = 0    1\nName: H5D_OPENS, dtype: int64.to_numpy
|
| darshan-test/python_runtime_tests.py:50: AssertionError
```

* there are many reasons I've opted not to use `pytest-mpi` at this
time, and gh-752 is only one of them

* if you'd like me to write an example test using C code, that should
be quite doable as well, we'd just have the `pytest` test compile
the source appropriately and then use the same approach; developing
this infrastructure was fairly time consuming though so I may want to
defer that to a follow-up if it is "ok"

* I'm not sure how many ranks we can safely use simultaneously
on the same free open source virtual machines, but this still
seems like a pretty big step up from "no runtime testing whatsoever"
in the CI

* this also opens up the possibility for i.e., using `gcov` to
produce reports of coverage of the C runtime code in CI, and so on..

* we may benefit from caching some steps.. building `h5py` from
source is time consuming in my local tests with `act`

### CODE REVIEW

**Comment by tylerjereddy** [comment]
One annoying thing here is that a few commands seem to require `sudo` in GitHub Actions CI proper, but not locally with `act`. Hopefully there is a way to have them both working symmetrically eventually, but for now I'll prioritize the CI version. There is a way to check if you are running locally with a variable so we could add that later maybe, to use/not use `sudo` depending on that.

**Comment by tylerjereddy** [comment]
Ok, CI seems to be running properly here now for the runtime testing, and picking up the current bug in the runtime. We could temporarily `xfail` the test until gh-751 is ready perhaps.

The 6 minute CI time is pretty reasonable as well, considering how much stuff gets compiled/downloaded.

File: .github/workflows/runtime_ci.yml
Comment (shanedsnyder) [inline .github/workflows/runtime_ci.yml:33]: you don't have to change this, but i confirmed you can work around the header location issue above by just adding `/usr/include/hdf5/openmpi/` to C_INCLUDE_PATH, i.e.,:

`export C_INCLUDE_PATH=$C_INCLUDE_PATH:/usr/include/hdf5/openmpi/`

in the "Install darshan-runtime step". Still an annoying additional step, but maybe less hack-y.

**Review by shanedsnyder — CHANGES_REQUESTED** [review]

**Comment by shanedsnyder** [comment]
I'm real late to the game here, but just had a look and think this looks great! I was able to get it working on my laptop using `act` and things appear to be working as expected.

I reviewed a couple of things, both really nitpicky, but at least the job ID env variable thing should probably be fixed. I'd probably be fine to merge after that.

A few additional comments:
- I'm fine with keeping the tests as failed for now rather than explicitly xfailing -- I plan to merge in the PR that fixes the underlying issue soon after this is merged, so that will avoid some churn in the repo.
- What exactly do you mean by caching some steps in the original PR description? Could that be used to cache some of the shared steps (like building darshan-util, installing pydarshan) from this CI workflow and the generic CI workflow we already have? Or did you have something else in mind? It would be nice if our different CI workflows could either reuse some of the same cached artifacts, or at the very least, could reuse some of the shared install scripts. That's probably out-of-scope here, I'm just asking to see if it's something worth following up on in an additional issue.

**Comment by shanedsnyder** [comment]
> One annoying thing here is that a few commands seem to require `sudo` in GitHub Actions CI proper, but not locally with `act`. Hopefully there is a way to have them both working symmetrically eventually, but for now I'll prioritize the CI version. There is a way to check if you are running locally with a variable so we could add that later maybe, to use/not use `sudo` depending on that.

I assume it's not really a big deal to let `act` run these commands locally using `sudo`, since it's all containerized? In fact, I kind of expected `sudo` would need to be used with `act` since, in my experience, you usually need to use `sudo` pretty liberally inside of containers?

**Comment by tylerjereddy** [comment]
I think I pushed in changes for the two requests from Shane.

- re: caching, yes I meant https://github.com/marketplace/actions/cache for example; a number of OSS projects use this to save time for stuff that gets repeated; not sure about cross-job cache, but certainly saving time on a given job; there's an example in SciPy (https://github.com/scipy/scipy/blob/main/.github/workflows/linux_meson.yml) though it is sufficiently complex that I'd probably agree we should do that in a separate PR
- re: `sudo` and `act`; ok, it works for me locally now too, so I guess that's not a problem anymore--I think my docker stuff got all messed up a few months ago when LANL changed some security settings..

**Review by shanedsnyder — APPROVED** [review]

**Pull request merged**
